Zendesk Mobile: create identity when it's actually needed.

ScoutHarris opened the pull request:
Fixes #9220 

To test:
- Fresh Install of the app.
- Log into a Self hosted site.
  - Open Me > Support.
  - Verify the three Support options work.
  - Find the `Zendesk identity created` log and verify the email and name match the site.
- Log into a WP account.
  - Open Me > Support.
  - Select one of the Support options.
  - Find the `Zendesk identity created` log and verify the email and name match the WP site.
- Log out of the WP account.
  - Open Me > Support.
  - Select one of the Support options.
  - Find the `Zendesk identity created` log and verify the email and name match the self hosted site.

To note: In the absence of a WP account, the app will use the _first_ self hosted site for Zendesk, regardless of additional self hosted sites added. Since this is written to User Defaults, a re-install is needed to clear it.

ScoutHarris commented:
@jleandroperez - I know you provided several options when we discussed this issue. However, they all depended on external factors. As I'm trying to keep Zendesk logic contained to `ZendeskUtils` as much as possible, I decided to let Zendesk create the identity when it needs it. Plus, that ensures the identity is created, instead of depending on it being triggered external.

As always, let me know what you think!

jleandroperez commented on WordPress/Classes/Utility/ZendeskUtils.swift:
```diff
@@ -37,78 +36,38 @@ import CoreTelephony
-        ZDKHelpCenter.presentOverview(presentInController, with: helpCenterContentModel)
+            let presentInController = ZendeskUtils.configureViewController(controller)
+            ZDKHelpCenter.presentOverview(presentInController, with: helpCenterContentModel)
+        }, failure: nil)
```
Few thoughts on alternatives!!

```
func showHelpCenterIfPossible(from controller:  UIViewController) {
    ZendeskUtils.createIdentity { success in {
          guard success else { 
              showFailureAlert(from: controller)
              return
          }

          showHelpCenter(from: controller)
    }
}

private func showHelpCenter(from controller: UIViewController) {
     // All of the `showHelpCenter` method's code, minus the Identity creation
}
```

- The main idea is to let the caller know that the method **may fail** (instead of silently failing).
- By updating the `createIdentify` signature to `private func createIdentity(completion: (Bool) -> Void)` you can *slightly* simplify the syntax. (Arguable point, definitely!!)
- This applies to the other public methods as well!

ScoutHarris replied:
Thanks @jleandroperez . I'll look at changing it. I hadn't quite figured out how I wanted to handle failures, but I can at least structure it well so it'll be easy to do/change in the future.

jleandroperez commented on WordPress/Classes/Utility/ZendeskUtils.swift:
```diff
@@ -279,7 +279,7 @@ private extension ZendeskUtils {
             let logData = try? Data(contentsOf: URL(fileURLWithPath: logFileInformation.filePath)),
             let logText = String.init(data: logData, encoding: .utf8) else {
-            return ""
+                return ""
```
Nipticky: Extra tab

jleandroperez replied:
Ahhhhhhh sorry about that!!

ScoutHarris commented:
@jleandroperez I liked your feedback. It looks much better now, thank you! 

I did the following:
- Added `IfPossible` to the public methods names.
- Changed `createIdentify` to only have a `completion` block.
- I added TODOs for showing errors. I have an item on my todo list to add error handling, so I'd like to address this later.

If you wouldn't mind another review, I'd greatly appreciate it!

ScoutHarris commented on WordPress/Classes/Utility/ZendeskUtils.swift:
```diff
@@ -279,7 +293,7 @@ private extension ZendeskUtils {
             let logFileInformation = fileLogger.logFileManager.sortedLogFileInfos.first,
             let logData = try? Data(contentsOf: URL(fileURLWithPath: logFileInformation.filePath)),
             let logText = String.init(data: logData, encoding: .utf8) else {
```
Thanks @jleandroperez . I have two outstanding PRs branched off of this one - I'll fix those last two items in one of those since they're so minor. I'm merging this thing! 😄

ScoutHarris commented:
Thanks @jleandroperez . I have two outstanding PRs branched off of this one - I'll fix those last two items in one of those since they're so minor. I'm merging this thing! 😄
